python/smqtk/algorithms/nn_index/mrpt.py

```diff
@@ -0,0 +1,451 @@
+            self._load_mrpt_model()
+
+        if (self._descr_cache_filepath and
+                osp.isfile(self._descr_cache_filepath)):
+            self.load_descr_cache(self._descr_cache_filepath)
+
+    def get_config(self):
+        return {
+            "index_filepath": self._index_filepath,
+            "parameters_filepath": self._index_param_filepath,
+            "descriptor_cache_filepath": self._descr_cache_filepath,
+            "random_seed": self._rand_seed,
+            "pickle_protocol": self._pickle_protocol,
+            "depth": self._depth,
+            "num_trees": self._num_trees,
+            "required_votes": self._required_votes,
+        }
+
+    def _has_model_files(self):
+        """
+        check if configured model files are configured and exist
+        """
+        return (self._index_filepath and osp.isfile(self._index_filepath) and
+                self._index_param_filepath and osp.isfile(self._index_param_filepath))
+
+    def build_index(self, descriptors):
+        """
+        Build the index over the descriptor data elements.
+
+        Subsequent calls to this method should rebuild the index, not add to
+        it, or raise an exception to as to protect the current index.
+
+        :raises ValueError: No data available in the given iterable.
+
+        :param descriptors: Iterable of descriptor elements to build index
+            over.
+        :type descriptors:
+            collections.Iterable[smqtk.representation.DescriptorElement]
+
+        """
+        super(MRPTNearestNeighborsIndex, self).build_index(descriptors)
+
+        self._log.info("Building new MRPT index")
+
+        self._log.debug("Storing descriptors")
+        self.init_descr_cache(descriptors, self._descr_cache_filepath,
+                              pickle_protocol=self._pickle_protocol)
+
+        self._log.debug("Accumulating descriptor vectors into matrix")
+        # XXX is an interval of 1.0 really a good idea?
+        pts_array = elements_to_matrix(self._descr_cache, report_interval=1.0)
+
+        self._log.debug('Building MRPT index')
+        self._build_multiple_trees(pts_array)
+        del pts_array
+
+        self._save_mrpt_model()
+
+        self._pid = multiprocessing.current_process().pid
+
+    def _build_multiple_trees(self, pts):
+        """
+        Build an MRPT structure for data pts
+        :param pts: The data. Each row is a datum.
+        :type pts: np.ndarray
+        """
+        n, d = pts.shape
+
+        # Get the Normal distribution RNG
+        rvs = stats.norm().rvs
+        # Start with no trees
+        self._trees = []
+        # 1/sqrt(depth) considered optimal for random projections
+        density = 1 / np.sqrt(self._depth)
+        for _ in range(self._num_trees):
+            # Each tree has a basis of sparse random projections
+            random_basis = random(d, self._depth, density=density, format=b"csc",
```

Comment:
Agreed on the first.
The `scipy.sparse` package converts matrices to CSR for (left) multiplication anyway, so you may as well use a CS{R,C} format to start out with. 

As it turns out, *right* multiplication of a dense matrix D by a sparse matrix S will actually do (S.T * D.T).T, and use left multiplication, so the transpose of D is probably the source of the copy on every iteration. (I'm glad I verified what I was claiming about multiplication.)